### PR TITLE
feat(squash): Add `--keep-emptied` flag

### LINKED ISSUE #3815: FR: `jj squash` should allow you to preserve the commit even if it was made empty
**Is your feature request related to a problem? Please describe.**
Squash is one command designed for two different semantic use cases. To be honest, this makes me very concerned, as it feels very much like the mistake git made with `checkout` where they needed to, later on, split it into `git switch` and `git restore`. 
* Move some files (technically, file diffs) from one commit into another
* Squash two commits into one, merging both files and commit metadata.

This presents issues when the files that I want to merge just happen to be all the files currently in the commit. For example, consider the following use case:

```
merge (empty)
|\
A B
```

1) I write code in the merge commit
1) I run `jj squash --from merge --to A`

I would want, in this scenario, to preserve everything about the merge commit:
* Commit description
* Change ID
* branch names
* topic names (once implemented)

However, in reality, my merge commit is abandoned, and it attempts to add the description "merge" to the parent commit (I can avoid this second part with `-u`). It then creates a new merge commit with an empty string as the description.

**Describe the solution you'd like**
I think that it's really nice to have a very clear distinction of "this command modifies commits. This command modifies files within commits". My personal preference would be to split `jj squash` into two commands.
* `jj squash` would work on whole commits. It would no longer support `--interactive`, or paths (if you wanted that, you would use `move` instead)
* `jj move` would work on files. It wouldn't touch anything related to commits.

However, maybe that's a controversial opinion (I'd be interested to see what others think to see whether it is a controversial one). I think a less extreme approach, if people don't like my first suggestion, would be:
* `jj squash` would get an option `--no-abandon` or something to that end, where it wouldn't abandon empty commits.
* `jj move` would be added to the default config as an alias to `["squash", --no-abandon"]`.

**Describe alternatives you've considered**
I've described multiple potential approaches above.

**Additional context**

### PR DESCRIPTION
**Pull request by matts1**
If the user was thinking in terms of files, instead of commits, then we shouldn't abandon the commit, as that can have side effects such as moving branch pointers.

Fixes #3815

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [N/A] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/squash.rs
Comment (mlcui-corp) [inline cli/src/commands/squash.rs:41]: rewording nit:

```suggestion
/// If `--interactive` is not set, no paths are provided, and `--keep-empty`
/// is not provided, the source will be abandoned.
```

Feel free to rephrase as necessary.

File: cli/src/commands/squash.rs
Comment (mlcui-corp) [inline cli/src/commands/squash.rs:68]: rewording nit to keep it consistent with other args:

```suggestion
    /// Interactively choose which parts to squash (implies --keep-empty)
```

File: cli/src/commands/squash.rs
Comment (mlcui-corp) [inline cli/src/commands/squash.rs:75]: rewording nit to keep it consistent with other args:

```suggestion
    /// Move only changes to these paths instead of all paths (implies --keep-empty)
```

Not too sure what the best way to handle the "double parentheses" is - feel free to rephrase, or ignore this comment.

File: cli/src/commands/squash.rs
Comment (mlcui-corp) [inline cli/src/commands/squash.rs:86]: nit: Is this intended to be included in this change?

**Review by mlcui-corp — COMMENTED** [review]
Thank you for implementing this! A few drive-by nits.

File: CHANGELOG.md
Comment (ilyagr) [inline CHANGELOG.md:40]: I slightly prefer `--keep-emptied`, actually.

File: cli/src/commands/squash.rs
Comment (ilyagr) [inline cli/src/commands/squash.rs:77]: I would make this more detailed:

"Keep revisions even if all of they are completely emptied by the squash operation"

File: CHANGELOG.md
Comment (ilyagr) [inline CHANGELOG.md:41]: I am not sure about this (still thinking a bit), but if we *are* changing the behavior in this way, it should be mentioned in "breaking changes".

File: cli/src/commands/squash.rs
Comment (ilyagr) [reply cli/src/commands/squash.rs:77]: Relatedely, I slightly prefer `--keep-emptied`.

Comment (ilyagr) [reply cli/src/commands/squash.rs:77]: We could also add a TODO to explain the similarity with `jj restore` in the common case of moving changes into a parent in a linear history (and that the operations are different in other cases).

File: CHANGELOG.md
Comment (ilyagr) [reply CHANGELOG.md:41]: I also wonder whether we should print a warning, either temporarily or permanently, if the commit is emptied but `--interactive` or paths are specified. Something like "The commit ... is now empty; use `jj abandon ...` to abandon it".

File: cli/src/commands/squash.rs
Comment (matts1) [reply cli/src/commands/squash.rs:77]: In retrospect, `--no-abandon` seems like a better choice to me. We are no longer doing anything conditional on whether the change is empty or not (it's purely based on the input the user provides now), so `--keep-empty` / `--keep-emptied` probably isn't a great choice..

Comment (ilyagr) [reply cli/src/commands/squash.rs]: It's a valid point that `--keep-emptied` does less, but I think that I ultimately disagree. As long as a regular `jj squash` with no options collapses two commits into one, that should be considered the "normal" behavior.

If the normal behavior was to keep both commits, I'd call the other behavior `--abandon-emptied`.

We could even have both `--keep-emptied` and `--abandon-emptied`, ultimately (instead of `--no-keep-emptied` or `--no-abandon-emptied`).

Another way of thinking about this: perhaps the main reason to have a `--no-blah` option should be to override a previously specified `--blah` option (in a config, say). This is most natural if `--no-blah` is the default behavior without any options. I just phrased this rule, so I may have missed some important exception, but it sounds good to me right now. :)

But I'll keep thinking about it, and perhaps others will chime in.

File: CHANGELOG.md
Comment (matts1) [reply CHANGELOG.md]: I added it to the breaking changes. I'm not opposed to adding a warning, but I'd like to hear other people's thoughts first.

Comment (ilyagr) [reply CHANGELOG.md]: I think this breaking change is a good idea, a person specifying paths is much more likely to want to `--keep-emptied`. My only worry is surprising people who relied on the old behavior or making their use-cases harder.

To avoid making use-cases harder, we could have an `--abandon-emptied` option, but in my mind the warning would make it unnecessary. It's easy enough to `jj abandon` the unnecessary commit; the only difficulty is that without a warning you'd have to check whether it is actually empty. (I'm also not sure how common such use-cases are, but I can imagine wanting both a squash and information on whether that nullified a commit when doing something like `jj squash cli/*` )

I don't think a warning is absolutely essential, though I like it more the more I think about it. OTOH, I know that @yuja and @martinvonz sometimes dislike such warnings.

File: cli/src/commands/squash.rs
Comment (matts1) [reply cli/src/commands/squash.rs]: Hmm, I'm on the fence about this. In my opinion:
* If we only have one flag, `--no-abandon` is a better name than `--keep-emptied`
  * `--keep-emptied` just adds confusion. Why are we saying this when whether we keep a commit is completely unrelated to whether it's empty?
* If we have two flags, `--keep-emptied` and `--abandon-emptied` is probably the better choice. They're clearly opposites.

I would probably be opposed to just implementing `--keep-emptied` right now (though I wouldn't be opposed to implementing both `--keep-emptied` and `--abandon-emptied`).

File: CHANGELOG.md
Comment (martinvonz) [inline CHANGELOG.md:12]: nit: I think people usually add new items last, unless there's a reason it belongs first (because it's more important than other changes or because it's related to the second item)

File: CHANGELOG.md
Comment (martinvonz) [inline CHANGELOG.md:1]: > Feat: Only abandon empty commits with squash if the user intended it.

nit: I think it's better to say "don't abandon empty commits on partial squash" or something like that. (It's hard to know what the user intended, so is it bug in this PR if an empty commit was abandoned when some user didn't intend it?)

File: cli/src/commands/move.rs
Comment (martinvonz) [inline cli/src/commands/move.rs:97]: I don't agree that the user clearly intended to leave the source commit behind. Maybe they just wanted to review the changes in their diff editor before confirming? I think it's a little unfortunate to lose that property, but maybe it's fine. I would prefer to not merge this just before the release later today.

FWIW, I view squash with paths or `-i` as a way of doing a partial squash; I don't think of it as "working on files" vs "working on  commits".

**Review by martinvonz — CHANGES_REQUESTED** [review]

File: cli/src/commands/squash.rs
Comment (matts1) [reply cli/src/commands/squash.rs]: Two caveats:
1) `jj squash .` only works from the root directory.
2) I'm also not sure how obvious it is to a user - I never even considered it myself until you brought it up.

I'm leaning towards  "implement both `--keep-emptied` and `--abandon-emptied`" (with the default being `--abandon-emptied`, and `--interactive` and paths both implying `--keep-emptied`).

Comment (yuja) [reply cli/src/commands/squash.rs]: If it's not so obvious, I think implying the default behavior is rather confusing. There's a valid use case of `jj squash --interactive --abandon-emptied`, and user might wonder why he has to specify `--abandon-emptied` only if `--interactive`.

Comment (matts1) [reply cli/src/commands/squash.rs]: Sorry, I think I was being ambiguous, which may have let to some confusion.

The thing that I'm not sure how  obvious it is to a user is the fact that you can use `jj squash .` to specify all files in the repository.

Comment (yuja) [reply cli/src/commands/squash.rs]: Ok, thanks for clarification.

I'm still not sure if `jj squash path` is clear sign to not abandon empty. To me, empty path arguments is usually equivalent to `.` or `root:"."`. Some commands like `jj log .` have subtle difference, but it's not super obvious either.

Comment (matts1) [reply cli/src/commands/squash.rs]: That was why my original proposal involved splitting it into two commands - `movediff` to work on diffs, and `squash` to work on commits. That would be my preference, but it got some pretty strong pushback, so I went with this instead.

Comment (yuja) [reply cli/src/commands/squash.rs]: Can't we simply add `--keep-emptied`/`--no-abandon` flag which is always off by default? User can add `diffmove` alias to turn it on. I don't think that'll provide the best UX, but it's less confusing than implying the default from path arguments.

Comment (ilyagr) [reply cli/src/commands/squash.rs]: Maybe `--never-abandon` is an OK name? It could be shortened to `--na`. Or perhaps we can come up with something better (I'd still prefer something without a negative in the name).

I still dislike `--no-abandon`, perhaps since it sounds like, without that option, something will definitely be abandoned. (this thought is definitely influenced by Matt's point above that I also link to below). But I could live with it if we can't agree on something else. 

In a world where we don't change the default behavior as Yuya suggested just above, I think `--keep-emptied` is also OK, but I don't insist on it.

In a world where we do change the default behavior, Matt's point from https://github.com/martinvonz/jj/pull/3830#discussion_r1626918110 gives me more pause.

(One other small argument against `--keep-emptied` is that @martinvonz pointed out on Discord that we might one day want something like `jj squash --keep`, that copies diffs instead of moving them. Though, perhaps that could be named `jj copydiff` or something.)

---------

From this discussion and https://github.com/martinvonz/jj/pull/3830#discussion_r1627391465, it sounds like changing the default behavior is in fact controversial. I kind of liked the idea in the other thread, but I'm getting cold feet after reading Martin and Yuya's points. Perhaps we should just add an option for now?

Comment (joyously) [reply cli/src/commands/squash.rs]: I'm not sure I follow all the ramifications of this, so I suggest to look at the big picture in terms of "changes" and "commits" and see if this is a special case. I've heard that you want to treat all commits the same way. Does that apply to all changes also? Is this a special case of `squash` or a `split`?

IMO, "changes" are the focus, and if you are squashing repeatedly, the emptied one is not a change to keep. Maybe it should be more clear if the flag is about the commit or the change.
I agree that most options should not be negative, but then don't most binary options get one of each, in order to override a config setting?

Comment (matts1) [reply cli/src/commands/squash.rs]: I've simplified this change down to only adding the flag, and not changing the default behaviour. I called it `--keep-emptied` for now, but I'm not particularly fussed on what we end up deciding on - I'll just go with whatever people want.

> Is this a special case of squash or a split

I don't think I would call it either. I'd call this an "amend" TBH. In my mind (and maybe other people don't think about this the same way), "amend" means "update an existing change", while squash means "2 changes become one".

> if you are squashing repeatedly, the emptied one is not a change to keep

While I'm sure this is true for your workflow, it's not an absolute truth of squashes. In general, if anyone ever wants to have a branch pointer pointing to a "working copy" commit, rather than a "committed" commit, it becomes a problem.

As an example of a use case where you might want that, when I work on a bunch of related, but not dependent commits, I may end up with a change graph looking like (this is simplified - in reality, I've had as many as 10 commits in parallel):
```mermaid
graph TD;
    foo-->server_change;
    foo-->client_change;
    server_change-->proto_change;
    client_change-->proto_change;
    proto_change-->main;
```

Here:
* `foo` is my "feature branch". Similar to any other squash-based workflow, it's roughly speaking my "working copy".
* `proto_change` is sent out for review with no dependencies
* `client_change` and `server_change` are sent out for review with `proto_change` as a diffbase

My workflow is:
1) Someone asks for changes on `client_change`
1) I write those changes in `foo`, and verify that the changes actually work with the server changes
1) I run `jj amend --to client_change` - currently, this will move my feature branch `foo` to `client_change`, which is clearly incorrect. 

This might seem like a weird workflow for someone working on github, given that github tends to think in terms of a PR and branches (which I believe is just a sequential stack of changes), but for someone working on gerrit, where you send individual changes for review, this merge commit approach allows you to parallelize your code reviews, and is what I believe to be the natural approach.

While I'm fully aware that there are ways to work around this, why should I need to work around it? Wanting to have a branch or a topic associated with a commit that may, at times, be empty, isn't invalid, even if it's not a thing everyone wants.

File: CHANGELOG.md
Comment (matts1) [reply CHANGELOG.md]: Marking as resolved since I'm no longer making this change a breaking one.

File: cli/src/commands/move.rs
Comment (matts1) [reply cli/src/commands/move.rs]: Changed to no longer infer behavior from flags, so I'll mark this a resolved.

File: cli/src/commands/squash.rs
Comment (yuja) [inline cli/src/commands/squash.rs:164]: nit: perhaps, this change is redundant. `.abandon` wouldn't be set if `!keep_empty`, and `Combine` means "combine abandoned".

File: CHANGELOG.md
Comment (yuja) [inline CHANGELOG.md:67]: nit: copy paste error?

**Review by yuja — APPROVED** [review]
Thanks for updating the patch. This looks good to me.

File: CHANGELOG.md
Comment (ilyagr) [reply CHANGELOG.md:1]: At this point, I would rename this to "`squash` Feat: `--keep-empty` option to not abandon empty commit", or something along those lines. I'd also update the commit description.

File: cli/tests/test_squash_command.rs
Comment (ilyagr) [inline cli/tests/test_squash_command.rs:277]: It would be nice to show empty commits in this log.

File: cli/tests/test_squash_command.rs
Comment (ilyagr) [inline cli/tests/test_squash_command.rs:291]: I'd add a comment: With `--keep-empty`, `b` remains even though it is now empty.

File: cli/tests/test_squash_command.rs
Comment (ilyagr) [inline cli/tests/test_squash_command.rs:301]: Optional: you could demo `squash` without `--keep-empty` here for comparison.

File: cli/src/commands/squash.rs
Comment (ilyagr) [reply cli/src/commands/squash.rs]: Thank you!

You currently called it `--keep-empty` (or perhaps it got renamed very recently). That's fine by me; I'm fine with either that or `--keep-emptied`.

**Review by ilyagr — APPROVED** [review]

File: CHANGELOG.md
Comment (ilyagr) [reply CHANGELOG.md:1]: Also, I'd rename the PR accordingly before merging it.

File: cli/tests/test_squash_command.rs
Comment (matts1) [reply cli/tests/test_squash_command.rs:301]: That's done in the test just above this one, so I won't bother.

**Comment by matts1** [comment]
One last thing I'd like to discuss before submitting is what the name of the flag would be. I'll wait for a few days before submitting for any feed back on  if people want the name changed.

Our options seem to be (feel free to add more if you have any ideas):
* `keep-emptied`
  * The most technically accurate mapping of behavior to name
  * Is "keep" going to be the term we *always* use in the future for "don't abandon". I'd like to be consistent about this.
* `keep-empty`
  * Technically, this would imply that we abandon *all* empty commits, which isn't correct in the one edge case where we squash two empty commits together.
* `no-abandon`
  * Would imply that the default is abandon (in reality, the default is abandon empty commits)
* `never-abandon`
  * Doesn't have the downsides of the implications of `--no-abandon`
  * A less common way of phrasing it
    * May be confusing, or intentionality may provoke thought and understanding of what it actually does.

File: CHANGELOG.md
Comment (ilyagr) [inline CHANGELOG.md:108]: I'll just leave this here as a reminder to sync up the commit description and contents before merging (after the option name is finalized). Currently, it's `--keep-emptied` in the name and `--keep-empty` in the contents.

Comment (martinvonz) [reply CHANGELOG.md:108]: And `jj rebase` has `--skip-empty` with "skip emptied" behavior. If we prefer the `-emptied` name because it's more correct, then we should also update `rebase` (in a separate PR).

**Review by martinvonz — APPROVED** [review]

**Pull request merged**
